File: src-model/model/ship/Ship.java
package model.ship;

import model.interfaces.Drawable;

import com.fasterxml.jackson.annotation.JsonIgnore;

import game.architecture.Entity;
import model.ship.parts.Control;
import model.ship.parts.Engine;
import model.ship.parts.Position;
import model.type.DrawableResolution;

public class Ship extends Entity implements Drawable {

    private Position position;
    private Engine engine;
    private Control control;
    private Mesh mesh;

    private double size;

    /* ========== PUBLIC ========== */
    public Ship(Position position, Engine engine, Mesh mesh, double size) {
        super(Ship.class);

        this.position = position;
        this.engine = engine;
        this.control = new Control();
        this.mesh = mesh;

        this.size = size;
        this.setResolution(DrawableResolution.VERY_FAR);
    }

    public Boolean visible() {
        return resolution != DrawableResolution.INVISIBLE;
    }

    /* ========== PRIVATE ========== */
    @JsonIgnore
    private DrawableResolution resolution;

    /* ========== PROPERTIES ========== */
    public Position getPosition() {
        return position;
    }

    public void setPosition(Position position) {
        this.position = position;
    }

    public Engine getEngine() {
        return engine;
    }

    public void setEngine(Engine engine) {
        this.engine = engine;
    }

    public Control getControl() {
        return control;
    }

    public void setControl(Control control) {
        this.control = control;
    }

    public Mesh getMesh() {
        return mesh;
    }

    public void setMesh(Mesh mesh) {
        this.mesh = mesh;
    }

    public double getSize() {
        return size;
    }

    public void setSize(double size) {
        this.size = size;
    }

    public DrawableResolution getResolution() {
        return resolution;
    }

    public void setResolution(DrawableResolution resolution) {
        this.resolution = resolution;
    }
}

File: src-model/model/ship/Ship.kt
package model.ship

import engine.math.Distance
import engine.math.Radius
import game.architecture.GameComponentContainer
import model.StarSystemObject
import model.celestials.CelestialBody
import model.celestials.Wormhole
import model.ship.parts.Control
import model.ship.parts.Engine
import model.ship.parts.Position
import model.system.StarSystem
import model.type.DrawableResolution

open class Ship(
        var position: Position,
        var engine: Engine,
        var mesh: Mesh?,
        val size: Float
) : StarSystemObject<Ship>(Ship::class.java, position.coords) {

    var resolution: DrawableResolution = DrawableResolution.VERY_CLOSE
    var control: Control = Control()
    var starSystem: StarSystem? = null

    fun visible(): Boolean? {
        return resolution !== DrawableResolution.INVISIBLE
    }

    fun updateDistance(body: CelestialBody) {
        val distance = distance(body)
        body.updateResolution(resolution(distance, body.radius()))
        checkCollisions(body, distance)
    }

    private fun resolution(distance: Distance, radius: Radius): DrawableResolution {
        return DrawableResolution.determineResolution(distance.div(radius))
    }

    private fun checkCollisions(body: CelestialBody, distance: Distance) {
        if (body is Wormhole) {
            if (distance.lessThan(body.radius())) {
                starSystem = GameComponentContainer.starSystemRepository.getById(body.systemToId)
            }
        }
    }
}